apollo-federation/src/link/federation_spec_definition.rs

```diff
@@ -114,6 +115,35 @@ impl FederationSpecDefinition {
         }
     }
 
+    pub(crate) fn federation_spec_to_string(&self) -> Result<String, FederationError> {
+        // Parse a simple schema to get an expanded subgraph
+        let subgraph = Subgraph::parse("S", "http://S", "type Query { hello: String }")?
+            .into_fed2_subgraph(self.version())?
+            .expand_links()?;
+        let mut result = String::new();
+
+        let directive_definitions = self.directive_specs();
+        for dir_def in directive_definitions {
+            if let Some(dir_def) = subgraph
+                .schema()
+                .schema()
+                .directive_definitions
+                .get(dir_def.name())
+            {
+                result.push_str(&dir_def.to_string());
+                result.push('\n');
```

Comment:
If you import `std::fmt::Write`, this can be simplified and you don't need to allocate a temporary string.
```suggestion
                _ = writeln!(result, "{dir_def}")
```
Note: `writeln!` returns a `Result`, but write a string should never fail (barring extreme situations), so we can just ignore it.